libs/sync/sync.go

```diff
@@ -134,13 +146,51 @@ func (s *Sync) notifyComplete(ctx context.Context, d diff) {
 func (s *Sync) RunOnce(ctx context.Context) error {
 	// tradeoff: doing portable monitoring only due to macOS max descriptor manual ulimit setting requirement
 	// https://github.com/gorakhargosh/watchdog/blob/master/src/watchdog/observers/kqueue.py#L394-L418
-	all, err := s.fileSet.All()
+	all := make([]fileset.File, 0)
+	gitFiles, err := s.fileSet.All()
 	if err != nil {
 		log.Errorf(ctx, "cannot list files: %s", err)
 		return err
 	}
+	all = append(all, gitFiles...)
+
+	include, err := s.includeFileSet.All()
+	if err != nil {
+		log.Errorf(ctx, "cannot list include files: %s", err)
+		return err
+	}
+
+	// Avoiding duplicates with Git tracked and include files
+	for _, i := range include {
+		if slices.ContainsFunc(all, func(a fileset.File) bool {
+			return a.Absolute == i.Absolute
+		}) {
+			continue
+		}
+
+		all = append(all, i)
+	}
+
+	all = append(all, include...)
+
+	exclude, err := s.excludeFileSet.All()
+	if err != nil {
+		log.Errorf(ctx, "cannot list exclude files: %s", err)
+		return err
+	}
+
+	files := make([]fileset.File, 0)
+	for _, f := range all {
+		if slices.ContainsFunc(exclude, func(a fileset.File) bool {
+			return a.Absolute == f.Absolute
+		}) {
+			continue
+		}
+
+		files = append(files, f)
+	}
```

Comment:
Please move this to a separate function, so it is separately testable.

I'm worried about the n^2 duplicate check. If you move this into a separate function, you could choose to maintain a map with files to make the dup check cheaper.

Reply:
@pietern I was leaning towards implementing a Set structure so we can reuse it later :)